src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -623,6 +628,28 @@ private BoundLambda ReallyInferReturnType(NamedTypeSymbol delegateType, Immutabl
             }
             lambdaSymbol.SetInferredReturnType(inferredReturnType.RefKind, returnType);
 
+            if (forErrorRecovery)
+            {
+                // When binding the lambda for error recovery, we bind one more time to permit the return expressions
+                // to be converted to the inferred return type so that no unconverted expressions remain in the bound tree.
+                diagnostics = DiagnosticBag.GetInstance();
```

Comment:
>diagnostics = DiagnosticBag.GetInstance(); [](start = 16, length = 42)

Leaking the bag from the pool?

Reply:
No, `diagnostics.ToReadOnlyAndFree()` appears below.

---
In reply to: [328736220](https://github.com/dotnet/roslyn/pull/38750#discussion_r328736220) [](ancestors = 328736220)